core/src/main/scala/kafka/server/ReplicaAlterDirThread.scala

```diff
@@ -0,0 +1,243 @@
+    val futureReplica = replicaMgr.getReplicaOrException(topicPartition, Request.FutureLocalReplicaId)
+    val records = partitionData.toRecords
+
+    if (fetchOffset != futureReplica.logEndOffset.messageOffset)
+      throw new RuntimeException("Offset mismatch for the future replica %s: fetched offset = %d, log end offset = %d.".format(
+        topicPartition, fetchOffset, futureReplica.logEndOffset.messageOffset))
+
+    // Append the leader's messages to the log
+    futureReplica.log.get.appendAsFollower(records)
+    futureReplica.highWatermark = new LogOffsetMetadata(partitionData.highWatermark)
+    futureReplica.maybeIncrementLogStartOffset(partitionData.logStartOffset)
+
+    if (replicaMgr.getPartition(topicPartition).get.maybeDeleteAndSwapFutureReplica())
+      removePartitions(Set(topicPartition))
+
+    if (quota.isThrottled(topicPartition))
+      quota.record(records.sizeInBytes)
+    brokerTopicStats.updateIntraBrokerBytesRate(topicPartition.topic, records.sizeInBytes)
+  }
+
+  def handleOffsetOutOfRange(topicPartition: TopicPartition): Long = {
+    val futureReplica = replicaMgr.getReplicaOrException(topicPartition, Request.FutureLocalReplicaId)
+    val replica = replicaMgr.getReplicaOrException(topicPartition)
+    val logEndOffset: Long = replica.logEndOffset.messageOffset
+
+    if (logEndOffset < futureReplica.logEndOffset.messageOffset) {
+      warn("Future replica for partition %s reset its fetch offset from %d to current replica's latest offset %d"
+        .format(topicPartition, futureReplica.logEndOffset.messageOffset, logEndOffset))
+      replicaMgr.logManager.truncateTo(Map(topicPartition -> logEndOffset), isFuture = true)
+      logEndOffset
+    } else {
+      val replicaStartOffset: Long = replica.logStartOffset
+      warn("Future replica for partition %s reset its fetch offset from %d to current replica's start offset %d"
+        .format(topicPartition, futureReplica.logEndOffset.messageOffset, replicaStartOffset))
+      val offsetToFetch = Math.max(replicaStartOffset, futureReplica.logEndOffset.messageOffset)
+      // Only truncate log when current leader's log start offset is greater than follower's log end offset.
+      if (replicaStartOffset > futureReplica.logEndOffset.messageOffset)
+        replicaMgr.logManager.truncateFullyAndStartAt(topicPartition, replicaStartOffset, isFuture = true)
+      offsetToFetch
+    }
+  }
+
+  def handlePartitionsWithErrors(partitions: Iterable[TopicPartition]) {
+    warn("Error while fetching data for future partitions %s".format(partitions.mkString(",")))
+  }
+
+  def buildLeaderEpochRequest(allPartitions: Seq[(TopicPartition, PartitionFetchState)]): Map[TopicPartition, Int] = {
+    val result = allPartitions
+      .filter { case (_, state) => state.isTruncatingLog }
+      .map { case (tp, _) => tp -> replicaMgr.getReplica(tp).get.epochs.get.latestEpoch }.toMap
```

Comment:
I maybe wrong here because I have not read through the KIP wiki of this epoch. The idea here is that ReplicaAlterDirThread will always copy&past the epoch from the current replica to the future replica of this partition. Thus it is safe to simply read the epoch and epoch offset from the current replica's cache. Did I miss something here?

Reply:
After a power failure, it's possible for the future replica to have more data than the current replica after recovery. New data could then be appended to the current replica before the future replica resumes replicating. You could have data like the following.

current replica:
offset:  0      1     2 
epoch:  1     1     2
value :  v1    v2   v4

future replica:
offset:  0      1     2 
epoch:  1     1     1
value :  v1    v2   v3

To reconcile the replicas, you want to use future replica's latest epoch (i.e. 1) to find the offset of the last offset in that epoch in the current replica so that the future replica can get rid of v3 and copy v4 over.

Reply:
Good point. I didn't consider the scenario that the future log may be ahead of the current log. I will updated the patch to use the latest epo of the future replica here.